services/payments/tests/unit/test_services_socketio.py

```diff
@@ -0,0 +1,262 @@
+    SOCKET_IO_PAYMENT_COMPLETED_EVENT,
+)
+from models_library.api_schemas_webserver.wallets import PaymentTransaction
+from models_library.users import GroupID
+from pydantic import parse_obj_as
+from pytest_mock import MockerFixture
+from pytest_simcore.helpers.rawdata_fakers import random_payment_transaction
+from pytest_simcore.helpers.typing_env import EnvVarsDict
+from pytest_simcore.helpers.utils_envs import setenvs_from_dict
+from servicelib.socketio_utils import cleanup_socketio_async_pubsub_manager
+from settings_library.rabbit import RabbitSettings
+from simcore_service_payments.models.db import PaymentsTransactionsDB
+from simcore_service_payments.services.rabbitmq import get_rabbitmq_settings
+from simcore_service_payments.services.socketio import notify_payment_completed
+from socketio import AsyncAioPikaManager, AsyncServer
+from tenacity import AsyncRetrying
+from tenacity.stop import stop_after_attempt
+from tenacity.wait import wait_fixed
+from yarl import URL
+
+pytest_simcore_core_services_selection = [
+    "rabbit",
+]
+pytest_simcore_ops_services_selection = []
+
+
+@pytest.fixture
+def app_environment(
+    monkeypatch: pytest.MonkeyPatch,
+    app_environment: EnvVarsDict,
+    with_disabled_postgres: None,
+    rabbit_env_vars_dict: EnvVarsDict,  # rabbitMQ settings from 'rabbit' service
+):
+    # set environs
+    monkeypatch.delenv("PAYMENTS_RABBITMQ", raising=False)
+
+    return setenvs_from_dict(
+        monkeypatch,
+        {
+            **app_environment,
+            **rabbit_env_vars_dict,
+        },
+    )
+
+
+@pytest.fixture
+def user_primary_group_id(faker: Faker) -> GroupID:
+    return parse_obj_as(GroupID, faker.pyint())
+
+
+@pytest.fixture
+async def socketio_server(app: FastAPI) -> AsyncIterable[AsyncServer]:
+    # Same configuration as simcore_service_webserver/socketio/server.py
+    settings: RabbitSettings = get_rabbitmq_settings(app)
+    server_manager = AsyncAioPikaManager(url=settings.dsn)
+
+    server = AsyncServer(
+        async_mode="aiohttp",
+        engineio_logger=True,
+        client_manager=server_manager,
+    )
+
+    yield server
+
+    await cleanup_socketio_async_pubsub_manager(server_manager)
+
+
+@pytest.fixture
+def socketio_server_events(
+    socketio_server: AsyncServer,
+    mocker: MockerFixture,
+    user_primary_group_id: GroupID,
+) -> dict[str, AsyncMock]:
+
+    user_room_name = f"{user_primary_group_id}"
+
+    # handlers
+    async def connect(sid: str, environ):
+        print("connecting", sid)
+        await socketio_server.enter_room(sid, user_room_name)
+
+    async def on_check(sid, data):
+        print("check", sid, data)
+
+    async def on_payment(sid, data):
+        print("payment", sid, parse_obj_as(PaymentTransaction, data))
+
+    async def disconnect(sid: str):
+        print("disconnecting", sid)
+        await socketio_server.leave_room(sid, user_room_name)
+
+    # spies
+    spy_connect = mocker.AsyncMock(wraps=connect)
```

Comment:
oh, nice one. did not know that was possible.
What is the difference with side_effect?

Reply:
@sanderegg I am spying... not "replacing" a function. i.e. the mock object here is not a replacement but a wrapper to capture the calls but at the end the actual function is executed . 

`pytest_mock`'s `mocker` fixture incorporate a `spy` method for convenience that can be used to spy on member functions. SEE https://pytest-mock.readthedocs.io/en/latest/usage.html#spy